pageserver/src/timelines.rs

```diff
@@ -0,0 +1,406 @@
+    ) -> Self {
+        let ancestor_timeline_id = timeline.get_ancestor_timeline_id();
+        let ancestor_lsn = if ancestor_timeline_id.is_some() {
+            Some(timeline.get_ancestor_lsn())
+        } else {
+            None
+        };
+
+        Self::Local {
+            timeline_id,
+            tenant_id,
+            last_record_lsn: timeline.get_last_record_lsn(),
+            prev_record_lsn: timeline.get_prev_record_lsn(),
+            ancestor_timeline_id,
+            ancestor_lsn,
+            disk_consistent_lsn: timeline.get_disk_consistent_lsn(),
+            current_logical_size: timeline.get_current_logical_size(),
+            current_logical_size_non_incremental: get_current_logical_size_non_incremental(
+                include_non_incremental_logical_size,
+                timeline,
+            ),
+        }
+    }
+
+    pub fn timeline_id(&self) -> ZTimelineId {
+        match *self {
+            TimelineInfo::Local { timeline_id, .. } => timeline_id,
+            TimelineInfo::Remote { timeline_id, .. } => timeline_id,
+        }
+    }
+}
+
+fn get_current_logical_size_non_incremental(
+    include_non_incremental_logical_size: bool,
+    timeline: &dyn Timeline,
+) -> Option<usize> {
+    if !include_non_incremental_logical_size {
+        return None;
+    }
+    match timeline.get_current_logical_size_non_incremental(timeline.get_last_record_lsn()) {
+        Ok(size) => Some(size),
+        Err(e) => {
+            error!("Failed to get non-incremental logical size: {:?}", e);
+            None
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PointInTime {
+    pub timeline_id: ZTimelineId,
+    pub lsn: Lsn,
+}
+
+pub fn init_pageserver(conf: &'static PageServerConf, create_tenant: Option<&str>) -> Result<()> {
+    // Initialize logger
+    // use true as daemonize parameter because otherwise we pollute zenith cli output with a few pages long output of info messages
+    let _log_file = logging::init(LOG_FILE_NAME, true)?;
+
+    // We don't use the real WAL redo manager, because we don't want to spawn the WAL redo
+    // process during repository initialization.
+    //
+    // FIXME: That caused trouble, because the WAL redo manager spawned a thread that launched
+    // initdb in the background, and it kept running even after the "zenith init" had exited.
+    // In tests, we started the  page server immediately after that, so that initdb was still
+    // running in the background, and we failed to run initdb again in the same directory. This
+    // has been solved for the rapid init+start case now, but the general race condition remains
+    // if you restart the server quickly. The WAL redo manager doesn't use a separate thread
+    // anymore, but I think that could still happen.
+    let dummy_redo_mgr = Arc::new(crate::walredo::DummyRedoManager {});
+
+    if let Some(tenantid) = create_tenant {
+        let tenantid = ZTenantId::from_str(tenantid)?;
```

Comment:
I've taken a stab at it and while it removes a bunch of code from pageserver, it brings the chicken and egg problem, as `create_tenant` is an http-based API method, while `init` phase is a CLI one and needs the `default_tenant_id` before writing the config now.

Starting and stopping the pageserver during `init` seems like an overkill IMO, and the CLI command for creating a tenant is kind of present already, so I think for now it can stay as is, while still confusing indeed.

We could add another command as a requirement to run for the user, but that does not seem right; we could implicitly update the `default_tenant` during `pageserver.start()` invocations, but that is quite odd either.
We could remove  the `pageserver init` command entirely and write a config, if it's absent, during `start`, but this is also not that convenient. Or `zenith init` could manage both configs at the same time, with `pageserver` completely lacking `init` again?

I'm fine with working on this in scope of this PR, just not sure which path is good.

Reply:
With init and start unification, there's a slight UX complication: you might want to `init` with `-c arg1=val1 -c..` once and then just run `pageserver start` always generally. But that's all worth a separate PR and discussion indeed.

What I've found useful to remove is the initial timeline creation during tenant creation in pageserver http api: this helps console and generally simpler: now the "initial timeline" concept is left only in cli `init`, actually, making it a bit more special and worth discussing further.

Reply:
After I've removed timeline creation from the tenant creation, https://github.com/zenithdb/zenith/issues/1294 became pretty simple to resolve by returning 200 on finding a repo in memory already, feel free to discuss that further in the issue.

Otherwise, I consider some improvement is done on that front, now only `init` has a notion of an initial timeline and maybe it's even fine for cli-oriented approach.